navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/instruction/DefaultInstructionEngine.java

```diff
@@ -0,0 +1,223 @@
+package com.mapbox.services.android.navigation.v5.instruction;
+
+import android.text.TextUtils;
+import android.util.SparseArray;
+
+import com.mapbox.services.android.navigation.v5.NavigationConstants;
+import com.mapbox.services.android.navigation.v5.RouteProgress;
+import com.mapbox.services.api.directions.v5.models.StepManeuver;
+import com.mapbox.services.api.utils.turf.TurfConstants;
+import com.mapbox.services.api.utils.turf.TurfHelpers;
+
+import java.text.DecimalFormat;
+import java.util.Locale;
+
+class DefaultInstructionEngine extends SparseArray<DefaultInstructionEngine.InstructionBuilder> {
+
+  private static final double MINIMUM_UPCOMING_STEP_DISTANCE = 15d;
+  private static final String DECIMAL_FORMAT = "###.#";
+  private static final String MILES_STRING_FORMAT = "%s miles";
+  private static final String FEET_STRING_FORMAT = "%s feet";
+  private static final String IN_STRING_FORMAT = "In %s %s";
+  private static final String THEN_STRING_FORMAT = "%s then %s";
+  private static final String THEN_IN_STRING_FORMAT = "%s then in %s %s";
+  private static final String CONTINUE_STRING_FORMAT = "Continue on %s for %s";
+
+  DefaultInstructionEngine() {
+    initDefaultBuilders();
+  }
+
+  private void initDefaultBuilders() {
```

Comment:
We should ticket out and work in a different branch how we could expose a way for developers to add their own instructions specific to their custom milestone. My understanding is right now if the user wants to add an instruction they'd need to implement their own instructionEngine or handle the string building inside `MilestoneEvent` callback.